bitforge/pubkey.py

```diff
@@ -1,84 +1,146 @@
+class InvalidEncoding(Error):
+    pass
 
-    class InvalidHex(Error, InvalidHex):
-        "The PublicKey string {string} is not valid hexadecimal"
 
+BasePublicKey = collections.namedtuple('PublicKey', [
+    'key',  # Elliptic curve public key
+    'network',  # Bitcoin-compatible network
+    'compressed',  # Whether the key should be serialized in compressed format
+])
 
-    def __new__(cls, pair, network = network.default, compressed = True):
-        if not utils.ecdsa.is_public_pair_valid(generator_secp256k1, pair):
-            raise PublicKey.InvalidPair(pair)
 
-        return super(PublicKey, cls).__new__(cls, pair, network, compressed)
+class PublicKey(BasePublicKey):
+    """Bitcoin public key."""
+
+    def __new__(cls, key, network=network.default, compressed=True):
+        """Create a Bitcoin public key from an EC public key."""
 
-    @staticmethod
-    def from_private_key(privkey):
-        pair = utils.public_pair_for_secret_exponent(
-            utils.generator_secp256k1, privkey.secret
-        )
+        return super(PublicKey, cls).__new__(cls, key, network, compressed)
 
-        # The constructor will validate the pair
-        return PublicKey(pair, privkey.network, privkey.compressed)
+    @classmethod
+    def from_point(cls, x, y, network=network.default, compressed=True, backend=default_backend()):
+        """Create a public key from its point coordinates.
+
+        A public key is a point on an elliptic curve, i.e. a pair (x, y) that
+        satisfies the curve equation.
+        """
+
+        public_numbers = ec.EllipticCurvePublicNumbers(x, y, network.curve)
 
-    @staticmethod
-    def from_bytes(bytes, network = network.default):
         try:
-            pair       = utils.encoding.sec_to_public_pair(bytes)
-            compressed = utils.encoding.is_sec_compressed(bytes)
-        except:
-            raise PublicKey.InvalidBinary(bytes)
+            key = public_numbers.public_key(backend)
+        except ValueError:
+            raise InvalidPoint()
+
+        return cls(key, network, compressed)
 
-        return PublicKey(pair, network, compressed)
+    @classmethod
+    def from_bytes(cls, data, network=network.default, backend=default_backend()):
+        """Create a public key from its raw binary encoding (in SEC1 format).
 
-    @staticmethod
-    def from_hex(string, network = network.default):
+        For more info on this format, see:
+
+        http://www.secg.org/sec1-v2.pdf, section 2.3.4
+        """
+
+        data = bytearray(data)
+
+        # A public key is a point (x, y) in the elliptic curve, and each
+        # coordinate is represented by a unsigned integer of key_size bytes
+        key_size = tools.elliptic_curve_key_size(network.curve)
+
+        # The first byte determines whether the key is compressed
         try:
-            bytes = decode_hex(string)
-        except InvalidHex:
-            raise PublicKey.InvalidHex(string)
+            prefix = data.pop(0)
```

Comment:
Is there a reason, or just preference? I went this way because you would need to do the pop later anyway.